packages/frontend/src/hooks/useChartColorConfig.ts

```diff
@@ -0,0 +1,97 @@
+import { FeatureFlags } from '@lightdash/common';
+import { useFeatureFlagEnabled } from 'posthog-js/react';
+import { useCallback } from 'react';
+
+/**
+ * Generic series-like object that contains the relevant portions used
+ * for color assignment. Requires the caller to unpack whatever series
+ * format it has into this compatible one.
+ */
+interface SeriesLike {
+    yField: string;
+    yPivotValues?: string[];
+}
+
+export const useChartColorConfig = ({
+    colorPalette,
+}: {
+    colorPalette: string[];
+}) => {
+    const useSharedColors = useFeatureFlagEnabled(
+        FeatureFlags.UseSharedColorAssignment,
+    );
+
+    /**
+     * Given the org's color palette, and an identifier, return the color palette value
+     * for said identifier.
+     *
+     * This works by hashing the identifier into an integer value, and then projecting
+     * that value into the org's color space - effectivelly getting a number from 0 to
+     * <number of colors in palette>.
+     *
+     * This is a straight-forward way to handle hashing dimensions into colors, with
+     * two major caveats:
+     *
+     * - We're no longer cycling over colors in the palette, potentially not following
+     *   an intentionally-designed best-neighbor-color approach.
+     * - We have no guarantee different dimensions won't be assigned the same color due
+     *   to the narrow color space - we can fix this by shifting colors aside when
+     *   compiling the chart config.
+     */
+    const calculateKeyColorAssignment = useCallback(
+        (identifier: string, colorsToAvoid: Set<string>) => {
+            const hashedValue = Math.abs(
+                identifier.split('').reduce(function (a, b) {
+                    a = (a << 5) - a + b.charCodeAt(0);
+                    return a & a;
+                }, 0),
+            );
+
+            const colorIdx = hashedValue % colorPalette.length;
+            const colorHex = colorPalette[colorIdx];
+
+            // Intersect colors we've already used with our color palette, and try to find the
+            // next available color on the list.
+            if (colorsToAvoid.has(colorHex)) {
```

Comment:
One thing to note, that I might tweak in a future PR - this branch has undefined behavior if the same dimension is seen more than once in the same chart+render cycle; I haven't seen it be an issue yet, and the outcome for now is the duplicate dimension will be assigned a different color, when it should instead reuse the previously calculated color.

Will keep an eye on it.